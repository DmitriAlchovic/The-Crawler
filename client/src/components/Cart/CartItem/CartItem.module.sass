.item-container
    cursor: pointer
    padding: 7px
    border: 1px solid #f3f3f3
    border-radius: 10px
    display: flex
    align-items: center

    .info-header
        position: relative
        width: 100%
        display: flex
        justify-content: space-between

        .info-container
            width: 60%

        .discount
            position: absolute
            background-color: #ffd480
            color: #ffff
            font-size: 12px
            height: 27px
            width: 78px
            display: flex
            justify-content: center
            align-items: center
            bottom: 0
            top: 0

        .product-img
            object-fit: contain
            width: 40%
            height: 154px

        .category
            margin-left: 10px
            font-size: 12px
            font-weight: 400
            color: #adadad

        .product-name
            margin-left: 10px
            font-size: 16px
            margin-top: 7px

        .quantity-container
            display: flex

            .quantity-changer
                border: none
                cursor: pointer
                background-color: #f3f3f3
                width: 3rem
                height: 3rem
                margin: 0 3px
                display: flex
                align-items: center
                justify-content: center
                border-radius: 10%

                .arrow-down
                    rotate: 180deg

        .sum-container
            font-weight: 600
            font-size: 1.2rem
