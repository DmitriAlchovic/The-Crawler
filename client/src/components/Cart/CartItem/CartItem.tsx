import React from 'react';
import { useAppDispatch } from '../../../hooks/redux';
import { changeQuantity } from '../../../store/reducers/cartSlice';
import { Product } from '../../../store/reducers/sliderSlice';
import ProductInfo from '../../ProductInfoCard/ProductInfo';
import arrow from '../../../assets/icon_arrow.svg';
import styles from './CartItem.module.sass';

interface CartItemProps {
  product: Product;
  itemIdx: number;
  item: { productId: number; quantity: number };
}

export default function CartItem({ product, itemIdx, item }: CartItemProps) {
  const { discountPercent } = product.discount;
  const { category } = product.subcategory;
  const { initialQuantity, currentQuantity } = product.inventory;
  const dispatch = useAppDispatch();

  return (
    <div className={styles['item-container']}>
      <div className={styles['info-header']}>
        {discountPercent && (
          <div className={styles.discount}>{`Save ${discountPercent}%`}</div>
        )}
        <img
          className={styles['product-img']}
          alt="no img"
          src={product.gallery}
        />
        <div className={styles['info-container']}>
          <div className={styles.category}>{category.name}</div>
          <div className={styles['product-name']}>{product.name}</div>
          <div className={styles['quantity-conatainer']}>
            <button
              className={styles['quantity-cahanger']}
              type="button"
              onClick={() => {
                dispatch(changeQuantity({ itemIdx, quantityItem: 1 }));
              }}
            >
              <img alt="up" src={arrow} />
            </button>
            <div className={styles['quantity-cahanger']}>{item?.quantity}</div>
            <button
              className={styles['quantity-cahanger']}
              onClick={() => {
                dispatch(changeQuantity({ itemIdx, quantityItem: -1 }));
              }}
              type="button"
            >
              <img className={styles['arrow-down']} alt="down" src={arrow} />
            </button>
          </div>
          <div className={styles['price-container']}>
            <ProductInfo
              initialQuantity={initialQuantity}
              price={product.price}
              currentQuantity={currentQuantity}
              discountPercent={discountPercent}
            />
          </div>
          {item && (
            <div className={styles['summ-container']}>
              {`Total: $${
                (product.price
                - (product.price * (discountPercent / 100))) * item.quantity
              }`}
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
