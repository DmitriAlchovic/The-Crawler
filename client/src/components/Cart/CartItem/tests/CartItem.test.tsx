import React from 'react';
import { render } from '@testing-library/react';
import * as reduxHooks from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import CartItem from '../CartItem';

jest.mock('react-redux');

describe('CartItem', () => {
  it('should create empty component', () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue('');
    const component = render(
      <BrowserRouter>
        <CartItem
          product={{
            productId: 0,
            name: '',
            price: 0,
            discount: {
              name: '',
              discountPercent: 0,
            },
            gallery: '',
            subcategory: {
              name: '',
              category: {
                name: '',
              },
            },
            inventory: {
              currentQuantity: 0,
              initialQuantity: 0,
            },
            desc: '',
            count: 0,
          }}
          itemIdx={0}
          item={{
            productId: 0,
            quantity: 0,
          }}
        />
      </BrowserRouter>,
    );

    expect(component).toMatchSnapshot();
  });
});
