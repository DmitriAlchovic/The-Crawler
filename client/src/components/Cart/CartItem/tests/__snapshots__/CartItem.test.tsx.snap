// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem should create empty component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="item-container"
      >
        <div
          class="info-header"
        >
          0
          <img
            alt="no img"
            class="product-img"
            src=""
          />
          <div
            class="info-container"
          >
            <div
              class="category"
            />
            <div
              class="product-name"
            />
            <div
              class="quantity-container"
            >
              <button
                class="disabled quantity-changer"
                type="button"
              >
                <img
                  alt="up"
                  src="icon_arrow.svg"
                />
              </button>
              <div
                class="quantity-changer"
              >
                0
              </div>
              <button
                class="quantity-changer"
                type="button"
              >
                <img
                  alt="down"
                  class="arrow-down"
                  src="icon_arrow.svg"
                />
              </button>
            </div>
            <div
              class="price-container"
            >
              <div
                class="info-body"
              >
                <div
                  class="discount-price"
                >
                  $0
                  0
                </div>
                0
              </div>
            </div>
            <div
              class="sum-container"
            >
              Total: $0
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="item-container"
    >
      <div
        class="info-header"
      >
        0
        <img
          alt="no img"
          class="product-img"
          src=""
        />
        <div
          class="info-container"
        >
          <div
            class="category"
          />
          <div
            class="product-name"
          />
          <div
            class="quantity-container"
          >
            <button
              class="disabled quantity-changer"
              type="button"
            >
              <img
                alt="up"
                src="icon_arrow.svg"
              />
            </button>
            <div
              class="quantity-changer"
            >
              0
            </div>
            <button
              class="quantity-changer"
              type="button"
            >
              <img
                alt="down"
                class="arrow-down"
                src="icon_arrow.svg"
              />
            </button>
          </div>
          <div
            class="price-container"
          >
            <div
              class="info-body"
            >
              <div
                class="discount-price"
              >
                $0
                0
              </div>
              0
            </div>
          </div>
          <div
            class="sum-container"
          >
            Total: $0
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
